doc/adr006-modules.lit.md

```diff
@@ -0,0 +1,553 @@
+      ...
+    end
+
+    finc --> inc
+    fdec --> dec
+    i1 --> dec_i
+    i2 --> inc_i
+    i3 --> dec_i
+    i4 --> inc_i
+    gN --> N
+    gx --> x
+```
+
+The interesting outcome of this approach is that we do not need a separate
+lookup table per module. Since we are using unique identifiers as keys, we
+could simply produce a flat table for all modules.
+
+### 4.2. Addressing the issue D3
+
+To address the need for treating variables of different instances as different
+objects, we propose to introduce a new flattening stage in the pipeline:
+
+```mermaid
+graph LR
+   repl["REPL"]
+   cli["CLI"] --> repl
+   cli --> loader["Source loader"] --> parser["Parser"]
+   repl --> loader
+   parser --> resolver["Name resolver"] --> analyzer
+   flattener["Flattener"]
+   analyzer --> flattener --> js
+   flattener --> apalache["Apalache transpiler"]
+
+   subgraph analyzer["Analyzer"]
+     types["Type inferrer"]
+     effects["Effect inferrer"]
+     updates["Updates checker"]
+     modes["Modes checker"]
+     types --> effects --> updates --> modes
+   end
+
+   subgraph runtime["JS runtime"]
+     js["JS transpiler"]
+     simulator["Simulator"]
+     testing["Unit testing"]
+
+     js --> simulator --> testing
+   end
+```
+
+The job of the flattener would be to clone all definitions of the instantiated
+module and tune them with the supplied expressions for the constants.
+Effectively, the flattener would transform the module `fixed` of our example
+into the following code:
+
+```bluespec generated/adr006instances.qnt +=
+module fixed {
+  // the instance general(N = 3)
+
+  // this is how `import fun.*` is transformed
+  import fun.* as I3
+
+  val I3::N = 3
+  var I3::x: int
+
+  action I3::init = { I3::x' = I3::N }
+  action I3::step = { I3::x' = I3::dec(x) }
+
+  // the instance general(N = 5)
+
+  // this is how `import fun.*` is transformed
+  import fun.* as I5
+
+  val I5::N = 3
+  var I5::x: int
+
+  action I5::init = { I5::x' = I5::N }
+  action I5::step = { I5::x' = I5::dec(x) }
+
+  // import I3.* is not allowed, see Section 4.3
+}
+```
+
+Note that the syntax `import fun.* as I3` is not supported yet. However,
+it does not seem too be hard to implement, if it's needed at all.
+
+### 4.3. Addressing the issue D4
+
+We propose the new syntax for declaring an instance:
+
+```antlr4
+  'include' name '(' [name = expr (',' name = expr)*] ')' 'as' name
+```
+
+The last name may be replaced with `_` to indicate that no prefix is required.
+
+Using this syntax, our example looks like follows:
+
+```bluespec
+module fixed {
+  include general(N = 3) as I3
+  include general(N = 5) as I5
+  
+  ...
+}
+```
+
+We believe that this syntax better reflects the nature of instances in Quint:
+
+ - All names of an instance are included into the namespace of the containing
+   module, possibly with a prefix, e.g., with `I3` or `I5`.
+
+ - Some names replaced with expressions, e.g., `N`.
+
+ - The behavior is different from `import`.
```

Comment:
Why? Isn't import a form of including but with no constant replacement? What is the motivation to differ them?

Reply:
`include` is almost like `import` but not only. It should also copy the definitions, when required. The semantics is similar but not exactly the same.